### ISSUE TITLE
Add support for Git submoodules

### ISSUE THREAD

**Issue by martinvonz** [martinvonz/jj#494]
## Description

Submodules are used frequently enough that I think we'll just have to support them.

**Comment by martinvonz** [comment]
I have been thinking about this a bit. One problem is how to handle working-copy commits.

We could create a new working-copy commit in each submodule when you check out a commit in the superproject. However, there are at least two problems with that. One is that it would make all submodules seem dirty because they point to a new commit, unless we somehow skip an empty working-copy commit when we report to the superproject what the submodule's commit is. Another problem is that we would have to scan the whole tree for submodules. That's particularly bad if you have a monorepo with millions of directories to walk; we don't want to do that every time you check out a commit.

Note that this problem is similar to what the Git project wants to do when you check out a branch. I think their plan is to check out the same branch in each submodule. The problem is smaller there because they presumably only need to do that for submodules that are within the sparse checkout, so that limits the number of directories to walk.

Another option is to somehow lazily create the new working-copy commits in the submodules only if you modify any files. Maybe we would also want to do it if you run `jj log` inside of them? If we do, then we'd probably still want to ignore those commits when we report the commit to the superproject so it doesn't seem like the submodule has been updated just because you ran `jj log` inside it.

I'm happy to hear if anyone has any thoughts (@chooglen?).

**Comment by wez** [comment]
FWIW, I run `git log` in a submodule to sanity check where I'm at and whether recent changes are present after a submodule update.  In that mode, I'm interested in the repo/commit history rather than the history of individual files.

I'd be fine with `jj log` not showing the submodule history in the working copy (honestly, I think it is confusing for the history to change based on the cwd in git anyway!) but it would be great if there was a way to explicitly review the history for a submodule when needed, eg: `jj log --submodule foo` or whatever would fit in consistently with the `jj` UX.

**Comment by chooglen** [comment]
This topic has turned out to be quite big, so I intend to break this up into smaller bits that we can discuss separately. There are a lot of things we'll have to figure out, but assuming we can't make big changes to the way Git implements submodules, I think these are the big questions:

* What is the mental model for submodules, and what does this mean for the UX?
* What is the on-disk submodules format?
* How will we snapshot changes in a submodule's content?
* How do we represent and handle submodule conflicts in the superproject?
* How will history rewriting the submodule be supported? Should even support it?

If these make sense, I'll open those issues separately.
Any other big questions I've missed?

**Comment by martinvonz** [comment]
> If these make sense, I'll open those issues (discussions?) separately.

Sounds good. (I'd make them issues, but discussions would also work.)

> What is the on-disk submodules format?

Do you mean for the working copy or the commit storage? Or both? I'm asking now only so you can decide if you want to create one or two separate issues for it (I'm fine with either).

**Comment by samueltardieu** [comment]
> Sounds good. (I'd make them issues, but discussions would also work.)

I think that is what "Projects" is for: you can make a project which groups all relevant issues, and see whether they are done, in progress or not started.

**Comment by chooglen** [comment]
> Do you mean for the working copy or the commit storage? Or both?

I think both, but since the two are linked (e.g. Git expects the .git file to point to the gitdir storing the commits), I'll open one discussion for that.

> I think that is what "Projects" is for: you can make a project which groups all relevant issues, and see whether they are done, in progress or not started.

That sounds good. Though AFAICT GitHub Projects don't live on a repo (you create a project on your own user account and then link it to one or many repos), and administering both the project and the repo sounds like a pain... I'll fiddle around with it a bit, and if it makes sense, I'll ask @martinvonz to create the project so that at least the repo and project have the same owner.

**Comment by rejuvenile** [comment]
I would vote for implementing git-subrepo. I like the model much better than the other git alternatives.

**Comment by senekor** [comment]
git-subrepo uses a different storage format than git submodules, so it's not relevant to this discussion.

**Comment by hemangandhi** [comment]
Hi!

What's the status of this? I wasn't sure where this is being discussed. From a quick grep and reading https://github.com/jj-vcs/jj/blob/main/docs/design/git-submodules.md, it looks like phase 1 is supported, but I'm not sure about the rest?

I also found https://github.com/jj-vcs/jj/blob/main/docs/design/git-submodule-storage.md and was wondering about its state: the document reads like a decision has been reached, but I can't understand the decision or why it's the one that was chosen? (I was actually thinking of experimenting with the [multi-backend alternative](https://github.com/jj-vcs/jj/blob/main/docs/design/git-submodule-storage.md#store-git-submodules-as-alternate-git-backends).)

**Comment by PhilipMetzger** [comment]
> Hi!
> 
> What's the status of this? I wasn't sure where this is being discussed. From a quick grep and reading https://github.com/jj-vcs/jj/blob/main/docs/design/git-submodules.md, it looks like phase 1 is supported, but I'm not sure about the rest?
> 
> I also found https://github.com/jj-vcs/jj/blob/main/docs/design/git-submodule-storage.md and was wondering about its state: the document reads like a decision has been reached, but I can't understand the decision or why it's the one that was chosen? (I was actually thinking of experimenting with the [multi-backend alternative](https://github.com/jj-vcs/jj/blob/main/docs/design/git-submodule-storage.md#store-git-submodules-as-alternate-git-backends).)

The Author of this design moved away from Source Control ~2 years ago and no-one picked this work up, it also isn't clear how it fits into the current `jj` so all in all there hasn't been any progress on this. Also Git Submodule support will probably require large design changes to Jujutsu why someone experienced with it should carry on with this work (its very similar to #80). TL;DR there hasn't been any progress here.

**Comment by hemangandhi** [comment]
Thank you for clarifying!

I'm sorry to hear that there's no progress and seemingly nobody taking this on.

If I do experiment with something like the [multi-backend alternative](https://github.com/jj-vcs/jj/blob/main/docs/design/git-submodule-storage.md#store-git-submodules-as-alternate-git-backends), can I share findings here?

> (its very similar to https://github.com/jj-vcs/jj/issues/80)

I see. (This is also another feature I was wondering about, but one I don't even understand where to start on.)

**Comment by PhilipMetzger** [comment]
> If I do experiment with something like the [multi-backend alternative](https://github.com/jj-vcs/jj/blob/main/docs/design/git-submodule-storage.md#store-git-submodules-as-alternate-git-backends), can I share findings here?

Yes, since this is the place about making the support for it possible, I encourage you to do so. There's also similar work underway for colocated workspaces (#8052) which share some similarities within the same design space. Also this post from Meta's Source Control team may also help here [Directory Branching](https://engineering.fb.com/2025/10/16/developer-tools/branching-in-a-sapling-monorepo).

**Comment by hemangandhi** [comment]
Hello. I've experimented a little, but have already started running into a few complications.

In particular, since my interest stemmed from https://github.com/jj-vcs/jj/issues/7490, I pursued making a backend that encapsulates other backends, in https://github.com/jj-vcs/jj/compare/main...hemangandhi:jj:heman/try-multi-git.

I reached the point where I was getting lost on basic user stories so am wondering: what sort of document would make sense to write and how should I proceed?

I think after writing the stories up, I can better determine if the experiment succeeded.
